python/cugraph-pyg/cugraph_pyg/examples/graph_sage_mg.py

```diff
@@ -0,0 +1,434 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from ogb.nodeproppred import NodePropPredDataset
+
+import time
+import argparse
+import gc
+
+import torch
+import numpy as np
+
+from torch_geometric.nn import CuGraphSAGEConv
+
+import torch.nn as nn
+import torch.nn.functional as F
+
+import torch.distributed as td
+import torch.multiprocessing as tmp
+from torch.nn.parallel import DistributedDataParallel as ddp
+
+from typing import List
+
+
+class CuGraphSAGE(nn.Module):
+    def __init__(self, in_channels, hidden_channels, out_channels, num_layers):
+        super().__init__()
+
+        self.convs = torch.nn.ModuleList()
+        self.convs.append(CuGraphSAGEConv(in_channels, hidden_channels))
+        for _ in range(num_layers - 1):
+            conv = CuGraphSAGEConv(hidden_channels, hidden_channels)
+            self.convs.append(conv)
+
+        self.lin = nn.Linear(hidden_channels, out_channels)
+
+    def forward(self, x, edge, size):
+        edge_csc = CuGraphSAGEConv.to_csc(edge, (size[0], size[0]))
+        for conv in self.convs:
+            x = conv(x, edge_csc)[: size[1]]
+            x = F.relu(x)
+            x = F.dropout(x, p=0.5)
+
+        return self.lin(x)
+
+
+def enable_cudf_spilling():
+    import cudf
+
+    cudf.set_option("spill", True)
+
+
+def init_pytorch_worker(rank, devices, manager_ip, manager_port) -> None:
+    import cupy
+    import rmm
+
+    device_id = devices[rank]
+
+    rmm.reinitialize(
+        devices=[device_id],
+        pool_allocator=False,
+    )
+
+    # torch.cuda.change_current_allocator(rmm.rmm_torch_allocator)
+    # cupy.cuda.set_allocator(rmm.rmm_cupy_allocator)
+
+    cupy.cuda.Device(device_id).use()
+    torch.cuda.set_device(device_id)
+
+    # Pytorch training worker initialization
+    dist_init_method = f"tcp://{manager_ip}:{manager_port}"
+
+    torch.distributed.init_process_group(
+        backend="nccl",
+        init_method=dist_init_method,
+        world_size=len(devices),
+        rank=rank,
+    )
+
+    # enable_cudf_spilling()
+
+
+def start_cugraph_dask_client(rank, dask_scheduler_file):
+    print(
+        "Connecting to dask... "
+        "(warning: this may take a while depending on your configuration)"
+    )
+    start_time_connect_dask = time.perf_counter_ns()
+    from distributed import Client
+    from cugraph.dask.comms import comms as Comms
+
+    client = Client(scheduler_file=dask_scheduler_file)
+    Comms.initialize(p2p=True)
+
+    end_time_connect_dask = time.perf_counter_ns()
+    print(
+        f"Successfully connected to dask on rank {rank}, took "
+        f"{(end_time_connect_dask - start_time_connect_dask) / 1e9:3.4f} s"
+    )
+    return client
+
+
+def stop_cugraph_dask_client():
+    from cugraph.dask.comms import comms as Comms
```

Comment:
I'm curious why this and other modules are imported inline, especially since some like this show up in multiple places.

Reply:
Importing cugraph will start a CUDA context prior to setting up rmm to use the right GPU